services/subscriber/service.go

```diff
@@ -494,3 +562,69 @@ func (b *balancewriter) Statistics(tags map[string]string) []models.Statistic {
 	}
 	return statistics
 }
+
+type dbrp struct {
+	db string
+	rp string
+}
+
+// subscriptionRouter has a mutex lock on the hot path for database writes - make sure that the lock is very tight.
+type subscriptionRouter struct {
+	mu            sync.RWMutex
+	ready         bool
+	m             map[dbrp][]*chanWriter
+	writeFailures *int64
+	Logger        *zap.Logger
+}
+
+func newSubscriptionRouter(statistics *Statistics) *subscriptionRouter {
+	return &subscriptionRouter{
+		ready:         true,
+		writeFailures: &statistics.WriteFailures,
+		Logger:        zap.NewNop(),
+	}
+}
+
+func (s *subscriptionRouter) Close() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.ready = false
+}
+
+func (s *subscriptionRouter) Send(request *coordinator.WritePointsRequest) {
+	// serialize points and put on writer
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if !s.ready {
+		return
+	}
+	writers := s.m[dbrp{
```

Comment:
Should we add the `, ok` 2nd return value here to guard against a missing key?

Reply:
That's covered by `len(writers) == 0` .

```
For a of map type M:
...
if the map is nil or does not contain such an entry, a[x] is the zero value for the element type of M
```

The zero value of a slice is nil, which is empty.